File: 1112-find-words-that-can-be-formed-by-characters/find-words-that-can-be-formed-by-characters.java
class Solution {
    public int countCharacters(String[] words, String chars) {

        int[] available = new int[26];

        for (char ch : chars.toCharArray()) {
            available[ch - 'a']++;
        }

        int ans = 0;

        for (String word : words) {

            int[] current = available.clone();
            boolean good = true;

            for (char ch : word.toCharArray()) {

                current[ch - 'a']--;

                if (current[ch - 'a'] < 0) {
                    good = false;
                    break;
                }
            }

            if (good) {
                ans += word.length();
            }
        }

        return ans;
    }
}

File: 1112-find-words-that-can-be-formed-by-characters/find-words-that-can-be-formed-by-characters.py
class Solution:
    def countCharacters(self, words, chars):
        freq = {}

        for ch in chars:
            freq[ch] = freq.get(ch, 0) + 1

        total = 0

        for word in words:
            current = freq.copy()
            good = True

            for ch in word:

                if current.get(ch, 0) == 0:
                    good = False
                    break

                current[ch] -= 1

            if good:
                total += len(word)

        return total